core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java

```diff
@@ -34,29 +49,29 @@
    @Override
    public void setContext(PrincipalRoleMapperContext context) {
       this.cacheManager = context.getCacheManager();
-      clusterRegistry = cacheManager.getGlobalComponentRegistry().getComponent(ClusterRegistry.class);
+      GlobalConfiguration globalConfiguration = cacheManager.getGlobalComponentRegistry().getGlobalConfiguration();
+      CacheMode cacheMode = globalConfiguration.isClustered() ? CacheMode.REPL_SYNC : CacheMode.LOCAL;
+      ConfigurationBuilder cfg = new ConfigurationBuilder();
+      cfg.clustering().cacheMode(cacheMode).sync().stateTransfer().fetchInMemoryState(true).awaitInitialTransfer(false);
+
+      InternalCacheRegistry internalCacheRegistry = cacheManager.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
+      internalCacheRegistry.registerInternalCache(CLUSTER_ROLE_MAPPER_CACHE, cfg.build(), EnumSet.of(InternalCacheRegistry.Flag.PERSISTENT));
    }
 
    public void grant(String roleName, String principalName) {
-      Set<String> roleSet = clusterRegistry.get(ClusterRoleMapper.class, principalName);
-      if (roleSet == null) {
-         roleSet = new HashSet<String>();
-      }
+      Set<String> roleSet = getClusterRoleMap().computeIfAbsent(principalName, n -> { return new HashSet<>(); });
```

Comment:
Whoops forgot computeIfAbsent passes the key as an argument, disregard my comment :-1: